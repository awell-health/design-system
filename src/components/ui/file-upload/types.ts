export type AcceptedFileTypes = '.jpg' | '.jpeg' | '.png' | '.pdf' | '.doc' | '.docx' | '.txt';
