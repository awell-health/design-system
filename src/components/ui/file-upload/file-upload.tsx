import { ChangeEvent, FC, useEffect, useRef, useState } from 'react';
import { Button } from '../button/button';
import { cn } from '../../../lib/utils';

interface Props {
  label?: string;
  onChange: (files: FileList) => void;
  error?: string;
  isMultiple?: boolean;
  accept?: string[];
  maxSizeMb?: number;
  onError?: (error: string) => void;
  translations?: {
    dropFilesHere: string;
    chooseFiles: string;
    someFilesExceedMaxSize: string;
    supportedFileTypes: string;
    maxFileSize: string;
  };
}

const FileUpload: FC<Props> = ({
  label,
  onChange,
  isMultiple = false,
  accept = ['.jpg', '.jpeg', '.png', '.pdf', '.doc', '.docx', '.txt'],
  maxSizeMb = 2, // Default max size to 2MB
  onError = undefined,
  translations = {
    dropFilesHere: 'Drop files here or',
    chooseFiles: 'Choose Files',
    someFilesExceedMaxSize: 'Some files exceed the maximum size of',
    supportedFileTypes: 'Supported file types:',
    maxFileSize: 'Maximum file size'
  }
}) => {
  const inputRef = useRef<HTMLInputElement>(null);
  const [isDragging, setIsDragging] = useState(false);
  const [error, setError] = useState<string | null>(null);

  useEffect(() => {
    if (error) {
      onError?.(error);
    }
  }, [error, onError]);

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    const files = e.target.files;

    if (files) {
      // validate files size
      const validFiles = Array.from(files).filter((file) => {
        return file.size <= maxSizeMb * 1024 * 1024;
      });

      if (validFiles.length !== files.length) {
        setError(`Some files exceed the maximum size of ${maxSizeMb}MB.`);
        return;
      }

      if (isMultiple) {
        onChange(validFiles as unknown as FileList);
      } else if (validFiles.length > 0) {
        onChange([validFiles[0]] as unknown as FileList);
      }
    }
  };

  const handleDragOver = (e: React.DragEvent<HTMLDivElement>) => {
    e.preventDefault();
    setIsDragging(true);
  };

  const handleDragLeave = (e: React.DragEvent<HTMLDivElement>) => {
    e.preventDefault();
    setIsDragging(false);
  };

  const handleDrop = (e: React.DragEvent<HTMLDivElement>) => {
    e.preventDefault();
    const files = e.dataTransfer.files;
    if (files) {
      const validFiles = Array.from(files).filter((file) => file.size <= maxSizeMb * 1024 * 1024);
      if (validFiles.length !== files.length) {
        setError(`Some files exceed the maximum size of ${maxSizeMb}MB.`);
      }
      if (validFiles.length > 0) {
        const event = new Event('change', { bubbles: true });
        Object.defineProperty(event, 'target', {
          value: { files: validFiles }
        });
        handleChange({ target: { files: validFiles } } as unknown as ChangeEvent<HTMLInputElement>);
      }
    }
    setIsDragging(false);
  };

  return (
    <div className='flex flex-col gap-2'>
      {label && <label className='text-sm font-medium text-gray-700'>{label}</label>}
      <input
        type='file'
        onChange={handleChange}
        className='hidden'
        ref={inputRef}
        multiple={isMultiple}
        accept={accept.join(',')}
      />
      <div
        className={cn(
          'flex flex-col items-center justify-center gap-4 border border-dashed border-gray-300 rounded-md p-4 relative w-full',
          {
            'border-blue-500': isDragging
          }
        )}
        onDragOver={handleDragOver}
        onDrop={handleDrop}
        onDragLeave={handleDragLeave}
      >
        <div>{translations.dropFilesHere}</div>
        <Button variant='secondaryBlue' size='sm' onClick={() => inputRef.current?.click()}>
          {translations.chooseFiles}
        </Button>
        <div className='text-sm text-gray-500 text-center'>
          {translations.maxFileSize} {maxSizeMb}MB <br />
          {translations.supportedFileTypes} {accept.join(', ')}
        </div>
        {isDragging && (
          <div className='absolute inset-0 bg-white/90 rounded-md'>
            <div className='flex items-center justify-center h-full'>
              <p className='text-gray-500 text-sm'>{translations.dropFilesHere}</p>
            </div>
          </div>
        )}
      </div>
      {error && <p className='text-red-500 text-sm'>{error}</p>}
    </div>
  );
};

export { FileUpload };
