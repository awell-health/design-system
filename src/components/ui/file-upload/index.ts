export { FileUpload } from './file-upload';
export type { AcceptedFileTypes } from './types';
