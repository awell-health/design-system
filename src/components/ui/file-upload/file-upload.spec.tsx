import React from 'react';
import { describe, expect, it, vi } from 'vitest';
import { fireEvent, render } from '@testing-library/react';
import { FileUpload } from './file-upload';

describe('FileList Component', () => {
  const subject = (props = {}) => {
    return render(<FileUpload onChange={() => {}} {...props} />);
  };

  it('should match snapshot', () => {
    const { container } = subject();
    expect(container).toMatchSnapshot();
  });

  it('triggers onChange when a file is uploaded', () => {
    const onChange = vi.fn();
    const { container } = subject({ onChange });
    const input = container.querySelector('input');

    if (input) {
      const file = new File(['(⌐□_□)'], 'chucknorris.png', { type: 'image/png' });
      Object.defineProperty(file, 'size', { value: 1024 * 1024 + 1 });

      fireEvent.change(input, { target: { files: [file] } });
      expect(onChange).toHaveBeenCalled();
    } else {
      throw new Error('Input not found');
    }
  });
});
