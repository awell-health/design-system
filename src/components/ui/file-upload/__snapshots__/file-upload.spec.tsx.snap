// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FileList Component > should match snapshot 1`] = `
<div>
  <div
    class="flex flex-col gap-2"
  >
    <input
      accept=".jpg,.jpeg,.png,.pdf,.doc,.docx,.txt"
      class="invisible"
      type="file"
    />
    <div
      class="flex flex-col items-center justify-center gap-4 border border-dashed border-gray-200 rounded-md p-4 relative w-full"
    >
      <div>
        Drop files here or
      </div>
      <button
        class="btn btn-sm rounded-md px-3.5 py-2 text-blue-700 bg-blue-50 border-blue-200 hover:bg-blue-100 hover:border-blue-300 active:bg-blue-100 active:border-blue-300 disabled:bg-blue-50 disabled:border-blue-200 disabled:text-blue-200 outline-blue-500"
      >
        Choose Files
      </button>
      <div
        class="text-sm text-gray-500 text-center"
      >
        Maximum file size
         
        2
        MB 
        <br />
        Supported file types:
         
        .jpg, .jpeg, .png, .pdf, .doc, .docx, .txt
      </div>
      <div
        class="absolute inset-0 bg-white/90 rounded-md transition-opacity duration-300 z-10 hidden"
      >
        <div
          class="flex items-center justify-center h-full"
        >
          <p
            class="text-gray-500 text-sm block"
          >
            Drop files here or
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
